feat: add a prettier pre-commit hook (#20)

KMK-Git opened the pull request:
Fixes #20 

Setup instructions:
- Run `npm install` to install all dependencies required for pre-commit hook to execute.

Implementation notes:
- Apart from prettier, I have used husky to implement the hook and lint-staged to execute prettier only on staged files instead of the full project before every commit.
- I have used an exact version of Prettier (2.7.1) based on the suggestion given in [Prettier docs](https://prettier.io/docs/en/install.html#summary). Other dependencies use `^` based on the convention used for other packages in the repo.
- https://github.com/KMK-Git/catchup/commit/6b368cea52deb58d4291979bf771a29462c9d7d7 covers all the changes made by Prettier to existing repo files. Since the formatting can change between Prettier versions, the output might be different from previous Prettier runs done with other versions.

tusharnankani commented on package.json:
```diff
@@ -5,7 +5,8 @@
 	"scripts": {
 		"start": "sh util/build.sh && nodemon index.js",
-		"alt-start": "./util/build.sh && nodemon index.js"
+		"alt-start": "./util/build.sh && nodemon index.js",
```
I don't understand why we have another `alt-start`.

Is there a reason why this has been added?

ref: [Difference b/w `sh filename.sh` and `./filename.sh`](https://askubuntu.com/questions/22910/what-is-the-difference-between-and-sh-to-run-a-script)
cc: @HarshKapadia2

KartikSoneji replied:
`./util/build.sh` won't work on Windows.
That being said, I'm not sure why we need `alt-start`, `sh util/build.sh` should work ok for unix as well.

KMK-Git replied:
To clarify and avoid any potential confusion, the `alt-start` script was already part of the repo and the diff here is because of the additional prepare script added to the repo.

KartikSoneji reviewed (changes requested):
Hi @KMK-Git , thanks for contributing to OTC catchup.
I see the original issue requests a pre-commit hook, but I think a less disruptive workflow would be to add a `npm run prettier` script that would run prettier automatically in the GitHub Actions workflow if one of the _supported_ files is changed by a commit.

Most of the content in the repo is in asciidoctor, which prettier doesn't support at the moment.
The html/js/css files that prettier does support are not changed frequently, so running prettier on every commit would add unnessary delay.

Also, maybe we should consider changing the remaining markdown files (readme, CoC and contributing) to adoc so they are consistent with the rest of the repo.

Related: [`NiklasPor/prettier-plugin-adoc`](https://github.com/NiklasPor/prettier-plugin-adoc), but it's not documented at all so I'm a bit hesitant to use it.

tusharnankani reviewed (commented):
Hi @KMK-Git 
Thank you so MUCH for your PR, it looks fine to me.

Although, I haven't used `lint-staged` or `husky` before. 
@KartikSoneji if you can take a look here.

I have a question.

> I have used an exact version of Prettier (2.7.1)

I understand. We have been using our custom settings here at [.prettierrc](https://github.com/OurTechCommunity/catchup/blob/main/.prettierrc)

https://github.com/OurTechCommunity/catchup/blob/f36fd9999787768098e4ec4b4472b3fc946c507f/.prettierrc#L15-L23

Also. most of the changes after adding the hook have been `tabWidths`s;  so I am not sure if I understand changing it just to default. 

So, I was wondering if we can inculcate `.prettierrc` settings in the hook?

tusharnankani commented:
> Also, maybe we should consider changing the remaining markdown files (readme, CoC and contributing) to adoc so they are consistent with the rest of the repo.

I did not understand this. Changing `md` to `adoc`?

KartikSoneji commented:
> > Also, maybe we should consider changing the remaining markdown files (readme, CoC and contributing) to adoc so they are consistent with the rest of the repo.
> 
> I did not understand this. Changing `md` to `adoc`?

Yes.

tusharnankani commented:
Why though?

KartikSoneji commented:
Everything else is in adoc, so I thought the repo should be consistant.
That will also allow us to publish the CoC and contributing instructions on the website.

KartikSoneji commented:
@KMK-Git gentle reminder.
Please let us know if you have any questions or would like some help.

KMK-Git commented:
@tusharnankani prettier is already using the config file, it doesn't need an explicit flag. [Check here the precedence order](https://prettier.io/docs/en/configuration.html)

@KartikSoneji Since the precommit hook uses lint-staged, it runs only on staged files. It doesn't run on all files in the project, so this should reduce any significant delays. Let me know next steps for this, should I:
1. Add a prettier step in build & build-and-deploy workflows?
2. Remove the pre-commit hook?
3. Any other changes.

KartikSoneji reviewed (changes requested):
Thanks for the update.

> It doesn't run on all files in the project, so this should reduce any significant delays.

Still, the precommit hook runs `npx lint-staged`, which in turn calls `prettier` before every commit adding a not-insignificant amount of delay especially on slower systems.

Since prettier will only be useful if one of the js scripts are changed, I think it makes more sense to add two npm scripts:
- `lint:check` that is run only in the `build` workflow and runs `prettier --check`
- `lint` that isn't run in the workflows, but is meant to fix linting issues with `prettier --write`

Please rebase your branch onto main and drop the commits for husky.

KMK-Git commented on .prettierignore:
```diff
@@ -0,0 +1,11 @@
+public/img/summary/
+public/html/summary/
+
+summary/autogenerated-combined-summary.adoc
```
The .prettierignore was just a copy of .gitignore since any file that is gitignored does not require linting. I removed the `summary/autogenerated-combined-summary.adoc` file since it doesn't need to be explicitly ignored for prettier, let me know if any other change is required in this file.

KMK-Git commented:
> Thanks for the update.
> 
> > It doesn't run on all files in the project, so this should reduce any significant delays.
> 
> Still, the precommit hook runs `npx lint-staged`, which in turn calls `prettier` before every commit adding a not-insignificant amount of delay especially on slower systems.
> 
> Since prettier will only be useful if one of the js scripts are changed, I think it makes more sense to add two npm scripts:
> 
>   * `lint:check` that is run only in the `build` workflow and runs `prettier --check`
> 
>   * `lint` that isn't run in the workflows, but is meant to fix linting issues with `prettier --write`
> 
> 
> Please rebase your branch onto main and drop the commits for husky.

@KartikSoneji 

Made the following changes:
1. Removed the husky commit.
2. Added `lint:check` and `lint` to `package.json` file.
3. The linting command cannot be added to the build stage since the asciidoctor image doesn't have npm installed. Instead of trying to install node/npm in that image, I created a separate lint stage which is executed first before build or build-and-deploy.
Example of failed lint check: https://github.com/KMK-Git/catchup/actions/runs/3349894537
Example of successful lint check: https://github.com/KMK-Git/catchup/actions/runs/3349904015 (deploy naturally fails in my repo because of missing token)

For the lint stage, I am using `node:18` image, but I see an ongoing conversation in #21 about whether a container image should be used in such scenarios. You can take a look at the `Initialize containers` step in the sample runs above to get a rough estimate of the extra time required.

KartikSoneji commented:
> > Thanks for the update.
> > > It doesn't run on all files in the project, so this should reduce any significant delays.
> > 
> > 
> > Still, the precommit hook runs `npx lint-staged`, which in turn calls `prettier` before every commit adding a not-insignificant amount of delay especially on slower systems.
> > Since prettier will only be useful if one of the js scripts are changed, I think it makes more sense to add two npm scripts:
> > 
> > * `lint:check` that is run only in the `build` workflow and runs `prettier --check`
> > * `lint` that isn't run in the workflows, but is meant to fix linting issues with `prettier --write`
> > 
> > Please rebase your branch onto main and drop the commits for husky.
> 
> @KartikSoneji
> 
> Made the following changes:
> 
> 1. Removed the husky commit.
> 2. Added `lint:check` and `lint` to `package.json` file.
> 3. The linting command cannot be added to the build stage since the asciidoctor image doesn't have npm installed. Instead of trying to install node/npm in that image, I created a separate lint stage which is executed first before build or build-and-deploy.
>    Example of failed lint check: https://github.com/KMK-Git/catchup/actions/runs/3349894537
>    Example of successful lint check: https://github.com/KMK-Git/catchup/actions/runs/3349904015 (deploy naturally fails in my repo because of missing token)
> 
> For the lint stage, I am using `node:18` image, but I see an ongoing conversation in #21 about whether a container image should be used in such scenarios. You can take a look at the `Initialize containers` step in the sample runs above to get a rough estimate of the extra time required.

Can you try running on the base image and see if that works?
I think nodejs should be available.

KMK-Git commented:
> > For the lint stage, I am using `node:18` image, but I see an ongoing conversation in #21 about whether a container image should be used in such scenarios. You can take a look at the `Initialize containers` step in the sample runs above to get a rough estimate of the extra time required.
> 
> Can you try running on the base image and see if that works? I think nodejs should be available.

https://github.com/KMK-Git/catchup/actions/runs/3351918379/jobs/5553692864 Sample run without container.

I also removed the `--only=dev` flag for npm install since it is no longer supported, adding it still installs all dependencies.

KMK-Git commented on .github/workflows/build-deploy-web-app.yaml:
```diff
@@ -27,7 +27,21 @@ on:
 
 jobs:
+    lint:
+        if: startsWith(github.ref, 'refs/pull/') || startsWith(github.ref, 'refs/heads/')
```
Removed

KMK-Git commented on .github/workflows/build-deploy-web-app.yaml:
```diff
@@ -27,7 +27,21 @@ on:
+              run: npm run lint:check
+
     build:
+        needs: lint
```
Removed dependency for `build`. Should I also remove it for `build-and-deploy` or keep it since deployed code should ideally be properly linted?

KartikSoneji replied:
No, the site should only be deployed if lint succedes.

KartikSoneji reviewed (approved):
LGTM
Thanks for contributing!

HarshKapadia2 commented:
Replying to [Kartik's comment](https://github.com/OurTechCommunity/catchup/pull/97?notification_referrer_id=NT_kwDOAv0WwLM0NjEzNjYyNzkyOjUwMTQwODY0#discussion_r1009700727):

> The `alt-start` script was added in https://github.com/OurTechCommunity/catchup/commit/3e5c7384d5b6641f3761298d90d1c27709a98b74.
> I don't remember why, it isn't used anywhere else in the repo either.
> @HarshKapadia2 do you see any reason to keep it around?

It was added to work in Gitpod, as the original `npm run start` command was not working. ([CONTRIBUTING.md#gitpod-setup](https://github.com/OurTechCommunity/catchup/blob/main/CONTRIBUTING.md#gitpod-setup))

KartikSoneji commented:
> Replying to [Kartik's comment](https://github.com/OurTechCommunity/catchup/pull/97?notification_referrer_id=NT_kwDOAv0WwLM0NjEzNjYyNzkyOjUwMTQwODY0#discussion_r1009700727):
> 
> > The `alt-start` script was added in [3e5c738](https://github.com/OurTechCommunity/catchup/commit/3e5c7384d5b6641f3761298d90d1c27709a98b74).
> > I don't remember why, it isn't used anywhere else in the repo either.
> > @HarshKapadia2 do you see any reason to keep it around?
> 
> It was added to work in Gitpod, as the original `npm run start` command was not working. ([CONTRIBUTING.md#gitpod-setup](https://github.com/OurTechCommunity/catchup/blob/main/CONTRIBUTING.md#gitpod-setup))

Let's discuss this here: #110
